Mathlib/Order/PrimeSeparator.lean

```diff
@@ -0,0 +1,155 @@
+  ⟨fun ⟨j, ⟨jJ, _, ha', bja'⟩⟩ => ⟨j, jJ, le_trans bja' (sup_le_sup_left ha' j)⟩,
+    fun ⟨j, hj, hbja⟩ => ⟨j, hj, a, le_refl a, hbja⟩⟩
+
+theorem prime_ideal_of_disjoint_filter_ideal (hFI : Disjoint (F : Set α) (I : Set α)) :
+    ∃ J : Ideal α, (IsPrime J) ∧ I ≤ J ∧ Disjoint (F : Set α) J := by
+
+  -- Let S be the set of proper ideals containing I and disjoint from F
+  set S : Set (Set α) := { J : Set α | IsIdeal J ∧ I ≤ J ∧ Disjoint (F : Set α) J }
+
+  -- Then I is in S...
+  have IinS : ↑I ∈ S := by
+    refine ⟨Order.Ideal.isIdeal I, by trivial⟩
+
+  -- ...and S contains upper bounds for any non-empty chains.
+  have chainub : ∀ c ⊆ S, IsChain (· ⊆ ·) c → c.Nonempty →  ∃ ub ∈ S, ∀ s ∈ c, s ⊆ ub := by
+    intros c hcS hcC hcNe
+    use sUnion c
+    refine ⟨?_, fun s hs ↦ le_sSup hs⟩
+    simp [S]
+    let ⟨J, hJ⟩ := hcNe
+    refine ⟨Order.isIdeal_sUnion_of_isChain (fun _ hJ ↦ (hcS hJ).1) hcC hcNe,
+            ⟨le_trans (hcS hJ).2.1 (le_sSup hJ), fun J hJ ↦ (hcS hJ).2.2⟩⟩
+
+  -- Thus, by Zorn's lemma, we can pick a maximal ideal J in S.
+  have zorn := zorn_subset_nonempty S chainub I IinS
+  have hJ := Exists.choose_spec zorn
+  set Jset := Exists.choose zorn
+  obtain ⟨⟨Jidl, IJ, JF⟩, ⟨_, Jmax⟩⟩ := hJ
+  set J := IsIdeal.toIdeal Jidl
+  use J
+  have IJ' : I ≤ J := IJ
+
+  clear chainub IinS
+
+  -- By construction, J contains I and is disjoint from F. It remains to prove that J is prime.
+  refine ⟨?_, ⟨IJ, JF⟩⟩
+
+  -- First note that J is proper: ⊤ ∈ F so ⊤ ∉ J because F and J are disjoint.
+  have Jpr : IsProper J := isProper_of_not_mem (Set.disjoint_left.1 JF F.top_mem)
+
+  -- Suppose that a₁ ∉ J, a₂ ∉ J. We need to prove that a₁ ⊔ a₂ ∉ J.
+  rw [isPrime_iff_mem_or_mem]
+  intros a₁ a₂
+  contrapose!
+  intro ⟨ha₁, ha₂⟩
+
+  -- Consider the ideals J₁, J₂ generated by J ∪ {a₁} and J ∪ {a₂}, respectively.
+  let J₁ := J ⊔ principal a₁
+  let J₂ := J ⊔ principal a₂
+
+  -- For each i, Jᵢ is an ideal that contains J, I and aᵢ, and is not equal to J.
+  have JsubJ₁ : J ≤ J₁ := le_sup_left
+  have JsubJ₂ : J ≤ J₂ := le_sup_left
+  have IJ₁ : I ≤ J₁ := le_trans IJ' le_sup_left
+  have IJ₂ : I ≤ J₂ := le_trans IJ' le_sup_left
+  have a₁J₁ : a₁ ∈ J₁ := mem_of_subset_of_mem (le_sup_right : _ ≤ J ⊔ _) mem_principal_self
+  have a₂J₂ : a₂ ∈ J₂ := mem_of_subset_of_mem (le_sup_right : _ ≤ J ⊔ _) mem_principal_self
+  have J₁J : ↑J₁ ≠ Jset := by refine ne_of_mem_of_not_mem' a₁J₁ ha₁
+  have J₂J : ↑J₂ ≠ Jset := by refine ne_of_mem_of_not_mem' a₂J₂ ha₂
+
+  -- Therefore, since J is maximal, we must have Jᵢ ∉ S.
+  have J₁S : ↑J₁ ∉ S := fun h => J₁J (Jmax J₁ h JsubJ₁)
+  have J₂S : ↑J₂ ∉ S := fun h => J₂J (Jmax J₂ h JsubJ₂)
+
+  -- Since Jᵢ is an ideal that contains I, we have that Jᵢ is not disjoint from F.
+  have J₁F : ¬ (Disjoint (F : Set α) J₁) := by
+    intro hdis
+    apply J₁S
+    simp only [le_eq_subset, mem_setOf_eq, SetLike.coe_subset_coe, S]
+    exact ⟨J₁.isIdeal, IJ₁, hdis⟩
```

Comment:
IJ₁ is used only here, couldn't you just insert the proof at this point?
Same for IJ₂ below, and some other functions that you introduce.

Reply:
Indeed, I'll try. Thanks